clang/lib/Driver/Driver.cpp

```diff
@@ -1117,6 +1117,24 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
         << SYCLHostCompilerOptions->getSpelling().split('=').first
         << "-fsycl-host-compiler";
 
+  Arg *SYCLUIHArg = C.getInputArgs().getLastArg(
+      options::OPT_fsycl_use_integration_headers,
+      options::OPT_fno_sycl_use_integration_headers);
+  if (SYCLUIHArg) {
+    // -f[no-]sycl-use-integration-headers cannot be used without
+    // -fsycl.  (-fintelfpga implies -fsycl).
+    if (!HasSYCL && !SYCLfpga)
+      Diag(clang::diag::err_drv_sycl_opt_requires_opt)
+          << "-f[no-]sycl-use-integration-headers";
+    // -fno-sycl-use-integration-headers cannot be used with
+    // -fsycl-host-compiler.
+    if (SYCLHostCompiler && SYCLUIHArg->getOption().matches(
+                                options::OPT_fno_sycl_use_integration_headers))
+      Diag(clang::diag::err_drv_option_conflict)
+          << SYCLHostCompiler->getSpelling().split('=').first
+          << "-fno-sycl-use-integration-headers";
+  }
```

Comment:
The use of `getLastArg()` above is good since it ensures that the `err_drv_sycl_opt_requires_opt` diagnostic is only issued when one of the options is explicitly used. I think we can simplify the check for conflict with the host compiler option a bit though and, while we're add it, perhaps future proof the code a bit against a change of default behavior. I suggest colocating the following with the declarations of `HasSYCL` and `HasValidSYCLRuntime`:
```
bool UseSYCLIntegrationHeaders =
    Args.hasFlag(options::OPT_fsycl_use_integration_headers,
                  options::OPT_fno_sycl_use_integration_headers,
                  true);
```
Then, break this diagnostic out of the enclosing `if` statement:
```suggestion
  }

  // -fno-sycl-use-integration-headers cannot be used with
  // -fsycl-host-compiler.
  if (SYCLHostCompiler && !UseSYCLIntegrationHeaders) {
    Diag(clang::diag::err_drv_option_conflict)
        << SYCLHostCompiler->getSpelling().split('=').first
        << "-fno-sycl-use-integration-headers";
  }
```

Reply:
Done.